File: snapo-app-mac/Snap-O/NetworkInspector/NetworkInspectorCopyExporter.swift
import AppKit
import Foundation

enum NetworkInspectorCopyExporter {
  static func copyURL(_ url: String) {
    setPasteboard(string: url)
  }

  static func copyCurl(for request: NetworkInspectorRequestViewModel) {
    let command = makeCurlCommand(for: request)
    setPasteboard(string: command)
  }

  private static func setPasteboard(string: String) {
    let pasteboard = NSPasteboard.general
    pasteboard.clearContents()
    pasteboard.setString(string, forType: .string)
  }

  private static func makeCurlCommand(for request: NetworkInspectorRequestViewModel) -> String {
    var warnings: [String] = []
    var parts: [String] = []

    parts.append("--request \(singleQuoted(request.method))")
    parts.append("--url \(singleQuoted(request.url))")

    for header in request.requestHeaders {
      parts.append("--header \(singleQuoted("\(header.name): \(header.value)"))")
    }

    if let bodyArgument = makeBodyArgument(for: request) {
      parts.append(bodyArgument.argument)
      warnings.append(contentsOf: bodyArgument.warnings)
    }

    let command = joinCurlParts(parts)
    guard !warnings.isEmpty else {
      return command
    }

    let warningLines = warnings.map { "# \($0)" }
    return (warningLines + [command]).joined(separator: "\n")
  }

  private static func makeBodyArgument(for request: NetworkInspectorRequestViewModel) -> (argument: String, warnings: [String])? {
    guard let body = request.requestBody else { return nil }

    var warnings: [String] = []

    if body.isPreview, body.truncatedBytes == nil {
      warnings.append("Request body is a preview – copied data may be incomplete")
    }

    if let truncated = body.truncatedBytes, truncated > 0 {
      warnings.append("Request body truncated by \(formatBytes(truncated)) – copied data may be incomplete")
    }

    let encoding = body.encoding?.lowercased()

    if encoding == "base64" {
      if let data = Data(base64Encoded: body.rawText) {
        let literal = makeBinaryLiteral(from: data)
        return ("--data-binary \(literal)", warnings)
      } else {
        warnings.append("Unable to decode base64 body – copied data uses raw text")
        return ("--data-binary \(singleQuoted(body.rawText))", warnings)
      }
    } else {
      return ("--data-binary \(singleQuoted(body.rawText))", warnings)
    }
  }

  private static func joinCurlParts(_ parts: [String]) -> String {
    guard !parts.isEmpty else { return "curl" }

    var remaining = parts
    let first = remaining.removeFirst()
    var firstLine = "curl \(first)"

    if !remaining.isEmpty {
      let second = remaining.removeFirst()
      firstLine += " \(second)"
    }

    guard !remaining.isEmpty else {
      return firstLine
    }

    var lines: [String] = []
    lines.append(firstLine + " \\")

    for (index, part) in remaining.enumerated() {
      let isLast = index == remaining.count - 1
      var line = "  \(part)"
      if !isLast {
        line += " \\"
      }
      lines.append(line)
    }

    return lines.joined(separator: "\n")
  }

  private static func singleQuoted(_ string: String) -> String {
    if string.isEmpty {
      return "''"
    }
    let escaped = string.replacingOccurrences(of: "'", with: "'\"'\"'")
    return "'\(escaped)'"
  }

  private static func makeBinaryLiteral(from data: Data) -> String {
    var result = "$'"
    result.reserveCapacity(data.count * 4)

    for byte in data {
      switch byte {
      case 0x07:
        result.append("\\a")
      case 0x08:
        result.append("\\b")
      case 0x09:
        result.append("\\t")
      case 0x0A:
        result.append("\\n")
      case 0x0B:
        result.append("\\v")
      case 0x0C:
        result.append("\\f")
      case 0x0D:
        result.append("\\r")
      case 0x5C:
        result.append("\\\\")
      case 0x27:
        result.append("\\'")
      default:
        if byte >= 0x20 && byte <= 0x7E {
          let scalar = Unicode.Scalar(byte)
          result.append(Character(scalar))
        } else {
          result.append(String(format: "\\x%02X", byte))
        }
      }
    }

    result.append("'")
    return result
  }

  private static func formatBytes(_ value: Int64) -> String {
    let formatter = ByteCountFormatter()
    formatter.countStyle = .binary
    return formatter.string(fromByteCount: value)
  }
}

File: snapo-desktop-compose/src/main/kotlin/com/openai/snapo/desktop/inspector/NetworkInspectorCopyExporter.kt
package com.openai.snapo.desktop.inspector

import java.awt.Toolkit
import java.awt.datatransfer.StringSelection
import java.util.Base64

object NetworkInspectorCopyExporter {
    fun copyUrl(url: String) = copyText(url)

    fun copyText(text: String) {
        val clipboard = Toolkit.getDefaultToolkit().systemClipboard
        clipboard.setContents(StringSelection(text), null)
    }

    fun copyCurl(request: NetworkInspectorRequestUiModel) {
        copyText(makeCurlCommand(request))
    }

    fun copyStreamEventsRaw(events: List<NetworkInspectorRequestUiModel.StreamEvent>) {
        val payload = events.joinToString(separator = "") { event ->
            val normalized = event.raw.replace(Regex("\\n+$"), "")
            normalized + "\n\n"
        }
        copyText(payload)
    }

    private fun makeCurlCommand(request: NetworkInspectorRequestUiModel): String {
        val warnings = mutableListOf<String>()
        val parts = mutableListOf<String>()

        parts += "--request ${singleQuoted(request.method)}"
        parts += "--url ${singleQuoted(request.url)}"

        for (header in request.requestHeaders) {
            parts += "--header ${singleQuoted("${header.name}: ${header.value}")}"
        }

        val bodyArg = makeBodyArgument(request, warnings)
        if (bodyArg != null) {
            parts += bodyArg
        }

        val command = joinCurlParts(parts)
        if (warnings.isEmpty()) return command

        val warningLines = warnings.map { "# $it" }
        return (warningLines + command).joinToString("\n")
    }

    private fun makeBodyArgument(
        request: NetworkInspectorRequestUiModel,
        warnings: MutableList<String>,
    ): String? {
        val body = request.requestBody ?: return null

        if (body.isPreview && body.truncatedBytes == null) {
            warnings += "Request body is a preview - copied data may be incomplete"
        }

        val truncated = body.truncatedBytes
        if (truncated != null && truncated > 0) {
            warnings += "Request body truncated by ${formatBytes(truncated)} - copied data may be incomplete"
        }

        val encoding = body.encoding?.lowercase()
        if (encoding == "base64") {
            val decoded = try {
                Base64.getDecoder().decode(body.rawText)
            } catch (_: Throwable) {
                null
            }

            return if (decoded != null) {
                "--data-binary ${makeBinaryLiteral(decoded)}"
            } else {
                warnings += "Unable to decode base64 body - copied data uses raw text"
                "--data-binary ${singleQuoted(body.rawText)}"
            }
        }

        return "--data-binary ${singleQuoted(body.rawText)}"
    }

    private fun joinCurlParts(parts: List<String>): String {
        if (parts.isEmpty()) return "curl"

        val remaining = parts.toMutableList()
        val first = remaining.removeFirst()
        var firstLine = "curl $first"

        if (remaining.isNotEmpty()) {
            val second = remaining.removeFirst()
            firstLine += " $second"
        }

        if (remaining.isEmpty()) return firstLine

        val lines = mutableListOf<String>()
        lines += "$firstLine \\"

        remaining.forEachIndexed { index, part ->
            val isLast = index == remaining.lastIndex
            lines += if (isLast) "  $part" else "  $part \\"
        }

        return lines.joinToString("\n")
    }

    private fun singleQuoted(value: String): String {
        if (value.isEmpty()) return "''"
        val escaped = value.replace("'", "'\"'\"'")
        return "'$escaped'"
    }

    private fun makeBinaryLiteral(data: ByteArray): String {
        val out = StringBuilder(data.size * 4)
        out.append("$'")

        for (b in data) {
            val byte = b.toInt() and 0xFF
            when (byte) {
                0x07 -> out.append("\\a")
                0x08 -> out.append("\\b")
                0x09 -> out.append("\\t")
                0x0A -> out.append("\\n")
                0x0B -> out.append("\\v")
                0x0C -> out.append("\\f")
                0x0D -> out.append("\\r")
                0x5C -> out.append("\\\\")
                0x27 -> out.append("\\'")
                in 0x20..0x7E -> out.append(byte.toChar())
                else -> out.append("\\x%02X".format(byte))
            }
        }

        out.append("'")
        return out.toString()
    }

    private fun formatBytes(byteCount: Long): String {
        // Keep it simple; ByteCountFormatter isn't available on the JVM stdlib.
        val kib = 1024.0
        val mib = kib * 1024.0
        val gib = mib * 1024.0

        val value = byteCount.toDouble()
        return when {
            value < kib -> "$byteCount B"
            value < mib -> "%.1f KiB".format(value / kib)
            value < gib -> "%.1f MiB".format(value / mib)
            else -> "%.1f GiB".format(value / gib)
        }
    }
}
